tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/lhlo_elemental_utils.cc

```diff
@@ -0,0 +1,302 @@
+
+    bool is_broadcast_dim = (it != broadcast_dimensions.end());
+    if (is_broadcast_dim) {
+      auto input_dim = std::distance(broadcast_dimensions.begin(), it);
+      auto static_dim_size =
+          operand_memref.getType().cast<MemRefType>().getShape()[input_dim];
+      if (static_dim_size == 1) {
+        // we know this dim is to be broadcasted at compile time
+        auto zero = b->create<ConstantOp>(
+            loc, b->getIndexType(), b->getIntegerAttr(b->getIndexType(), 0));
+        input_index.push_back(zero);
+      } else if (static_dim_size == ShapedType::kDynamicSize) {
+        // we are not sure if this dim is to be broadcasted at compile time
+        auto dim_size = b->create<DimOp>(loc, operand_memref, input_dim);
+        auto one = b->create<ConstantOp>(
+            loc, b->getIndexType(), b->getIntegerAttr(b->getIndexType(), 1));
+        auto zero = b->create<ConstantOp>(
+            loc, b->getIndexType(), b->getIntegerAttr(b->getIndexType(), 0));
+        auto dim_size_is_1 =
+            b->create<CmpIOp>(loc, CmpIPredicate::eq, dim_size, one);
+        input_index.push_back(b->create<mlir::SelectOp>(
+            loc, dim_size_is_1, zero, output_index[dim]));
+      } else {
+        // we know this dim is not to be broadcasted at compile time
+        input_index.push_back(output_index[dim]);
+      }
+    }
+  }
+
+  if (!check_cache) {
+    auto rank = operand_memref.getType().dyn_cast<MemRefType>().getRank();
+    return (rank > 0) ? b->create<LoadOp>(loc, operand_memref, input_index)
+                      : b->create<LoadOp>(loc, operand_memref, ValueRange());
+  }
+  return createLoadOrUseCachedValue(loc, b, operand_memref, input_index,
+                                    b->saveInsertionPoint());
+}
+
+}  // namespace
+
+template <>
+Value elementalLower<lmhlo::DynamicBroadcastInDimOp>(OpBuilder* b, Location loc,
+                                                     Operation* op,
+                                                     ValueRange output_index,
+                                                     bool check_cache) {
+  return elementalLowerImplForBroadcastInDimOps<lmhlo::DynamicBroadcastInDimOp>(
+      b, loc, op, output_index, check_cache);
+}
+
+template <>
+Value elementalLower<lmhlo::BroadcastInDimOp>(OpBuilder* b, Location loc,
+                                              Operation* op,
+                                              ValueRange output_index,
+                                              bool check_cache) {
+  return elementalLowerImplForBroadcastInDimOps<lmhlo::BroadcastInDimOp>(
+      b, loc, op, output_index, check_cache);
+}
+
+scf::ForOp createLoopAndSetInsPt(OpBuilder& b, Location loc, Value& var,
+                                 Value lb, Value ub, Value step,
+                                 ArrayRef<Value> init_values) {
+  auto for_op = b.create<scf::ForOp>(loc, lb, ub, step, init_values);
+  for_op.getBody()->clear();
```

Comment:
Isn't the body always empty on creation?

Reply:
For a very early version of the loop dialect, the builder of loop::ForOp will create an empty yield terminator by default, which is not always expected when there are initial values. I checked the codes of MLIR and this problem has already been fixed. So yes I can remove the code here. Done.

Reply:
For a very early version of the loop dialect, the builder of loop::ForOp will create an empty yield terminator by default, which is not always expected when there are initial values. I checked the codes of MLIR and this problem has already been fixed. So yes I can remove the code here. Done.